### PR TITLE
Fix required permissions in subscription query editor

### PR DESCRIPTION
**Pull request by Cloud11PL**
## What type of PR is this?
- [ ] 💅 Refactor
- [ ] 🌟 Feature
- [ ] 🔥 Bug Fix
- [ ] 🔩 Maintenance
- [ ] 🛠 Workflow CI/CD changes

## Related Issues or Documents

- closes #

## Usage Instructions, Screenshots, Recordings
CleanShot 2024-07-25 at 12 29 35
## Have you written tests?
- [ ] Yes!
- [ ] No... here is why: _Writing tests are mandatory, please replace this text with why test are not included in this PR_

### CODE REVIEW

**Review by krzysztofzuraw — APPROVED** [review]

File: src/custom-apps/components/PermissionAlert/utils.ts [hunk 1]
```diff
@@ -1,47 +1,32 @@
 // @ts-strict-ignore
```
Comment (poulch) [inline src/custom-apps/components/PermissionAlert/utils.ts:1]: suggestion: If you rearrange a whole file would be good to remove it and fix types accordingly to rules

**Review by poulch — DISMISSED** [review]

File: src/custom-apps/components/PermissionAlert/utils.ts [hunk 2]
```diff
@@ -1,47 +1,53 @@
...
-      types {
-        kind
-        name
-        description
-        fields(includeDeprecated: false) {
-          name
-          description
-        }
-      }
+const tokenTree = (token: Token | undefined, tokens: string[] = []): string[] => {
```
Comment (poulch) [inline src/custom-apps/components/PermissionAlert/utils.ts:32]: question: What does this function do? IMO it is not a token tree it is more like flattened token tree

File: src/custom-apps/components/PermissionAlert/utils.ts [hunk 3]
```diff
@@ -1,47 +1,53 @@
...
+  if (token && token.next) {
+    if (typeof token.value === "string") {
+      tokens.push(token.value);
     }
```
Comment (poulch) [inline src/custom-apps/components/PermissionAlert/utils.ts:36]: question: shouldn't you push token value no matter if token has next?

File: src/custom-apps/components/PermissionAlert/utils.ts [hunk 4]
```diff
@@ -50,92 +56,119 @@ const uniq = <T>(value: T, index: number, self: T[]) => self.indexOf(value) ===
...
+
+    const _permissions = Object.keys(descriptions).reduce((acc, key) => {
+      const { isOneOfRequired, permissions } = extractPermissions(descriptions[key]);
+
+      if (permissions.length > 0) {
+        acc[key] = { isOneOfRequired, permissions };
+      }
+
+      return acc;
+    }, {} as SubscriptionQueryPermission);
```
Comment (poulch) [inline src/custom-apps/components/PermissionAlert/utils.ts:166]: suggestion: That could be also be wrapped into separate function and `extractPermissionsFromQuery ` would look like this:

```suggestion
const extractPermissionsFromQuery = (
  query: string,
  introspectionQuery: IntrospectionQuery,
): SubscriptionQueryPermission => {
    try {
      const schema = buildClientSchema(introspectionQuery);
      const descriptions = getDescriptionsFromQuery(query, schema);
      
      return getAllPermission(descriptions) // You name it better 
    } catch (error) {
      // explicit silent
    }
}
```

**Review by poulch — APPROVED** [review]

File: src/custom-apps/components/PermissionAlert/utils.ts (hunk 3 above)
Comment (Cloud11PL) [reply src/custom-apps/components/PermissionAlert/utils.ts:36]: Doesn't matter here that much, I did that to filter out empty values

**Review by poulch — APPROVED** [review]

**Review by krzysztofzuraw — APPROVED** [review]

**Pull request merged**